### PR TITLE
Hotfix/upack existing package

### PR DESCRIPTION
**Pull request by mihairatoiu**
**Task name**: UniversalPackages

**Description**: On existing package publish fail with error instead of success.

**Documentation changes required:** (Y/N) N

**Added unit tests:** (Y/N) N

**Attached related issue:** Y
https://github.com/microsoft/azure-pipelines-tasks/issues/14487
**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/UniversalPackagesV0/universalpublish.ts [hunk 1]
```diff
@@ -149,6 +149,12 @@ export async function run(artifactToolPath: string): Promise<void> {
             retries++; 
         }
 
+        if(publishStatus === packageAlreadyExistsError){
```
Comment (zhenghao104) [inline Tasks/UniversalPackagesV0/universalpublish.ts:152]: Thanks for catching the bug. This if check can be moved to line 118 before the while loop

Comment (zhenghao104) [inline Tasks/UniversalPackagesV0/universalpublish.ts:152]:
```suggestion
        if(publishStatus === packageAlreadyExistsError && versionRadio === "custom"){
```

File: Tasks/UniversalPackagesV0/universalpublish.ts [hunk 2]
```diff
@@ -149,6 +149,12 @@ export async function run(artifactToolPath: string): Promise<void> {
...
+            throw new Error(tl.loc("Error_PublishingRetry",
```
Comment (zhenghao104) [inline Tasks/UniversalPackagesV0/universalpublish.ts:153]: I dont think we need a new string for the error, the debug log will show package already exist error, we can throw it the same way as the[ old implementation ](https://github.com/microsoft/azure-pipelines-tasks/blob/92ae148a02b1a174cc212222f9d92b07dc536ea6/Tasks/UniversalPackagesV0/universalpublish.ts#L204)

File: Tasks/UniversalPackagesV0/universalpublish.ts (hunk 1 above)
Comment (zhenghao104) [reply Tasks/UniversalPackagesV0/universalpublish.ts:152]: and we can remove `versionRadio !== "custom"` on line 120

File: Tasks/UniversalPackagesV0/task.loc.json [hunk 3]
```diff
@@ -425,6 +425,7 @@
     "UnknownFeedType": "ms-resource:loc.messages.UnknownFeedType",
     "Error_NoSourceSpecifiedForPublish": "ms-resource:loc.messages.Error_NoSourceSpecifiedForPublish",
     "Error_NoSourceSpecifiedForDownload": "ms-resource:loc.messages.Error_NoSourceSpecifiedForDownload",
+    "Error_PublishingRetry": "ms-resource:loc.messages.Error_PublishingRetry",
```
Comment (zhenghao104) [inline Tasks/UniversalPackagesV0/task.loc.json:428]: Not needed, see comment below

File: Tasks/UniversalPackagesV0/task.json [hunk 4]
```diff
@@ -425,6 +425,7 @@
         "UnknownFeedType": "Unknown feed type '%s'",
         "Error_NoSourceSpecifiedForPublish": "No external source was specified for publish",
         "Error_NoSourceSpecifiedForDownload": "No external source was specified for download",
+        "Error_PublishingRetry": "Package %s, version: %s already exists.",
```
Comment (zhenghao104) [inline Tasks/UniversalPackagesV0/task.json:428]: Not needed, see comment below

**Review by zhenghao104 — CHANGES_REQUESTED** [review]

File: Tasks/UniversalPackagesV0/universalpublish.ts (hunk 2 above)
Comment (mihairatoiu) [reply Tasks/UniversalPackagesV0/universalpublish.ts:153]: If I change the Error to Error_UnexpectedErrorArtifactTool I will need to give the error message that is not passed to the outside scope from publishPackageUsingArtifactTool. Adding this new error solves this issue and also gives more insight.

File: Tasks/UniversalPackagesV0/universalpublish.ts (hunk 1 above)
Comment (mihairatoiu) [reply Tasks/UniversalPackagesV0/universalpublish.ts:152]: moving this before the while will not cover all cases. In the case of non-custom version, after 1 retry(maybe uncommon, but possible) no error will be thrown if the package was not published

Comment (zhenghao104) [reply Tasks/UniversalPackagesV0/universalpublish.ts:152]: I see, agree.

File: Tasks/UniversalPackagesV0/universalpublish.ts (hunk 2 above)
Comment (zhenghao104) [reply Tasks/UniversalPackagesV0/universalpublish.ts:153]: The old string already contains the exit code from artifacttool and the error message of `package already exist`

These are the[ exit code](https://dev.azure.com/mseng/AzureDevOps/_git/ArtifactTool?path=%2Fsrc%2FArtifactTool%2FErrorCode.cs) that get returned from artifact tool. Throwing using the current string we have should capture both the exit code as well as the type of error. Look [here](https://github.com/microsoft/azure-pipelines-tasks/blob/a009bb0cd3a2b17b83770826bbf37c0d187a90bd/Tasks/UniversalPackagesV0/universalpublish.ts#L210).

Comment (mihairatoiu) [reply Tasks/UniversalPackagesV0/universalpublish.ts]: removed extra error template

File: Tasks/UniversalPackagesV0/task.loc.json (hunk 3 above)
Comment (mihairatoiu) [reply Tasks/UniversalPackagesV0/task.loc.json]: Removed

**Comment by tldavies** [comment]
@zhenghao104 @aasim @phil-hodgson @satbai  @mihairatoiu  - I was wondering if we can get an update on when this will get merged. As it is cause confusion in our publish builds.

**Comment by mihairatoiu** [comment]
I'm also waiting on the reviewers

**Comment by thisisbrianstewart** [comment]
Looking for this fix to be merged on our end as well, happy to help out in any way I can, but looks like it is still pending reviews.

**Review by zhenghao104 — APPROVED** [review]

**Comment by zhenghao104** [comment]
Approved, only the task version needs to be updated to match the current sprint.

**Comment by mihairatoiu** [comment]
Sure, I have updated the task version

**Comment by mihairatoiu** [comment]
Is there something else needed to do? I see that It's waiting for status on the pipelines for a couple of days...

**Comment by zhenghao104** [comment]
You may have to update this branch with the base branch to kick off those pipeline.

**Comment by mihairatoiu** [comment]
I did that but I see it didn't work and it's now again out-of-date with the base branch

**Comment by mihairatoiu** [comment]
@zhenghao104 can you please retrigger the azure pipelines to run? I have updated to master but it didn't trigger.

**Comment by zhenghao104** [comment]
Retriggered

**Comment by zhenghao104** [comment]
@mihairatoiu  @tldavies  @thisisbrianstewart Can this be merged?

**Review by thisisbrianstewart — APPROVED** [review]
Looks good, no comments to add.  Did notice another issue during the review that I will open separately -- this PR correctly checks the error code with ===, but previous code used ==, and we should favor strict equality unless we expect that return type to mutate.

**Comment by thisisbrianstewart** [comment]
> @mihairatoiu @tldavies @thisisbrianstewart Can this be merged?

Reviewed and good to go on my end.

**Comment by zhenghao104** [comment]
@thisisbrianstewart  I am not able to merge from my end. @mihairatoiu

**Comment by thisisbrianstewart** [comment]
@zhenghao104 it likely requires someone who is a code owner to merge it.  @aasim @phil-hodgson @satbai could you review this PR and assist?

**Comment by mihairatoiu** [comment]
Sorry, I don't have the option to merge it. The code owners should be able to.

**Comment by aasim** [comment]
@mihairatoiu - can you please merge this branch with master

**Comment by mihairatoiu** [comment]
I have updated to current version of master and increased the version to current sprint

**Pull request merged**